piptools/writer.py

```diff
@@ -103,7 +103,7 @@ def write_header(self) -> Iterator[str]:
     def write_index_options(self) -> Iterator[str]:
         if self.emit_index_url:
             for index, index_url in enumerate(dedup(self.index_urls)):
-                if index_url.rstrip("/") == self.default_index_url:
+                if index == 0 and index_url.rstrip("/") == self.default_index_url:
```

Comment:
Oh, nice catch! Have you checked by the way how pip combines `--index-url` and `--extra-index-urls`? We should keep in sync with that logic.

Reply:
Pip just [defaults `--index-url` to PyPI](https://github.com/pypa/pip/blob/751cbb72f5f7c3f0b98e2c1ff50fb61abde41acb/src/pip/_internal/cli/cmdoptions.py#L324) and [joins both index and extra index urls](https://github.com/pypa/pip/blob/751cbb72f5f7c3f0b98e2c1ff50fb61abde41acb/src/pip/_internal/req/req_file.py#L244). It then resolves dependencies from the index in the list order.

This is also how I hit this. If you set your private index as index-url and then add pypi as extra-index-url to make sure the private index is checked first, the requirements.txt pip-tools produces currently is not installable because it doesn't contain the extra-url option for pypi and so all public packages aren't found. If you add it manually to the file everything works as expected.

Reply:
Thanks for the analysis!